docs/design/offline-cascading-deletes/README.md

```diff
@@ -0,0 +1,414 @@
+                "id": <UUID v1>,
+                "partitionKey": <UUID of changed document>
+              }
+              ```
+
+        5. Commit transaction
+
+### Dispatch
+
+Events will be dispatched from the outbox to a message broker / queue through
+changed data capture (CDC) using Debezium. See [Meadowlark Streaming and
+Downstream Data Stores](https://techdocs.ed-fi.org/x/UgWgBw) for more
+information.
+
+### Offline Process - Event Handler / Stream Processor
+
+There will be a monitoring service / function (details forthcoming) that detects
+the `key-update` event and then takes action to fix dependencies. Naive approach:
+
+1. For each document that references the `event.documentUuid` ("child document"):
+   1. For each component of `event.newIdentity`: replace child document's value
+      for matching key.
+   2. Calculate the child document's new Meadowlark ID and update the document.
+   3. Create a `key-update` event _for the child document_, so that any
+      "grandchildren" of the original change will also receive the update.
+   4. Create a `document-update` event for the child document, for updating
+      OpenSearch (details forthcoming).
+
+Why is this naive? Because it does not account for (at least) two conflict
+situations.
+
+1. Multiple updates to the key that are handled in the wrong order, and
+2. Updates to both a parent key and a child document.
+
+Note that the first problem, as described below, could occur with the ODS/API.
+The second problem is unique to Meadowlark; it is not possible on in a
+relational database.
+
+#### Multiple updates
+
+What if there are multiple updates to the key, and the events are handled in the
+wrong order? Then the wrong update would end up "winning". This scenario could
+occur when two different threads or processes accept HTTP requests. Conceivably,
+an API client sends an update, then quickly sends a second update with a
+correction. If the first one is processed slowly, then the erroneous update
+might be the second in the event stream.
+
+```mermaid
+sequenceDiagram
+    Client->>+API-1: PUT /section/1
+    Client->>+API-2: PUT /section/1
+    API-2->>+DB: Save section & key-update
+    API-1->>+DB: Save section & key-update
+    DB->>+Events: section (latest) & key-update
+    API-2->>+DB: Save section & key-update
+    DB->>+Events: section (older) & key-update
+
+```
+
+The problem needs to be resolved upstream of the events - before recording the
+change in the primary data store. The likely solution is to have the API modify
+the document _immediately on receipt_, enriching with a Unix timestamp property
+`lastModifiedAt`. So long as the server times are in sync, then the API should
+avoid writing an update if the stored value has a larger `lastModifiedAt`. To be
+on the safe side, use similar logic in the event handling. This is a variation
+of optimistic locking. It may be helpful to think of this timestamp as an
+"identity version number".
```

Comment:
Shoutout to "identity version number" mentioned by Brad in a ticket comment.

Reply:
I like this. Easier that managing a global "next version" counter somewhere.